### ISSUE TITLE
--cookie-refresh does not actually refresh with OIDC

### ISSUE THREAD

**Issue by rassie** [oauth2-proxy/oauth2-proxy#1077]
## Expected Behavior

Judging by the docs, `--cookie-refresh` should refresh the access tokens after indicated time interval (and the token cookie, if sessions are not used). While we are waiting on https://github.com/oauth2-proxy/oauth2-proxy/issues/856, this should allow transparent refreshing of the access token via the refresh token.

## Current Behavior

Currently, with OIDC provider the token refresh only happens after the session has expired. 

In `refreshSessionIfNeeded`, refresh period is correctly checked for:

https://github.com/oauth2-proxy/oauth2-proxy/blob/4fa607f173a01d591e65913ae56423dd55175b56/pkg/middleware/stored_session.go#L117-L120

If the refresh period has expired, provider is asked to perform a session refresh

https://github.com/oauth2-proxy/oauth2-proxy/blob/4fa607f173a01d591e65913ae56423dd55175b56/pkg/middleware/stored_session.go#L123

However, in OIDC's `RefreshSessionIfNeeded`, the refresh token is only redeemed unless a session is empty, refresh token is missing or current session has expired:

https://github.com/oauth2-proxy/oauth2-proxy/blob/4fa607f173a01d591e65913ae56423dd55175b56/providers/oidc.go#L119-L121

## Possible Solution

`oidc.RefreshSessionIfNeeded` should take the `RefreshPeriod` into account.

## Steps to Reproduce (for bugs)

1.  Setup an OIDC IDP with access token lifetime of 120 seconds
2.  Setup a minimal API behind `oauth2-proxy`, with `--cookie-refresh` set to `60s`.
3.  Create a minimal webpage which triggers the API endpoint every 15 seconds
4.  Observe logs, which should show a normal authenticated request for the first four requests, every consequent request should trigger a `Refreshing 1m1.005128459s old session cookie for` message but not actually refresh the token (can be checked by adding logging output to `oidc.RefreshSessionIfNeeded`) and after two minutes there should be a 50/50 chance of the token actually refreshing or refresh failing because of expired session. 

## Context

I'm trying to get "silent refresh" to work and fail.

## Your Environment

- Version used: Latest git snapshot.

**Comment by rassie** [comment]
In https://github.com/oauth2-proxy/oauth2-proxy/issues/856 the introduction says:
> Currently, when we are passed the cookie-refresh TTL, we only refresh the session if the Access or ID Token is expired (depending on which the provider uses for the ExpiresOn field).

Which is basically the description of this issue. Even though I would very much like #856 to be implemented, it's probably some time away and I think actually using `RefreshPeriod` at all the right places would do some good in the meantime.

**Comment by rassie** [comment]
It seems this is actually a duplicate of #793, with an attemped fix in https://github.com/oauth2-proxy/oauth2-proxy/pull/1041, folded back into #856. Closing and waiting for #856 :)
